contrib/babelfishpg_tsql/sql/sys_procedures.sql

```diff
@@ -271,3 +271,227 @@ GRANT EXECUTE on PROCEDURE sys.sp_babelfish_volatility(IN sys.varchar, IN sys.va
+      IF var_object_id IS NULL THEN
+        RAISE EXCEPTION 'Object is invalid. Extended properties are not permitted on ''%'', or the object does not exist.', coalesce(var_object_name, 'object specified');
+        RETURN;
+      END IF;
+
+      IF "@level2type" IS NULL THEN
+        var_type := "@level1type";
+      ELSE
+        var_type := "@level1type" || ' ' || "@level2type";
+        var_minor_name := "@level2name";
+        var_object_name := var_schema_name || '.' || var_major_name || '.' || var_minor_name;
+        IF "@level1type" IN ('TABLE') AND "@level2type" IN ('COLUMN') AND NOT EXISTS (SELECT * FROM pg_attribute WHERE attrelid = var_object_id AND attname = var_minor_name COLLATE "C") THEN
+          RAISE EXCEPTION 'Object is invalid. Extended properties are not permitted on ''%'', or the object does not exist.', coalesce(var_object_name, 'object specified');
+          RETURN;
+        END IF;
+      END IF;
+    END IF;
+  END IF;
+
+  SELECT current_user INTO var_user;
+  PERFORM sys.babelfish_set_role('sysadmin');
+
+  procedure_name = LOWER(procedure_name);
+  IF NOT EXISTS (SELECT * FROM sys.babelfish_extended_properties WHERE dbid = var_dbid AND schema_name = var_schema_name COLLATE "C" AND major_name = var_major_name COLLATE "C" AND minor_name = var_minor_name COLLATE "C" AND type = var_type COLLATE "C" AND name = "@name") THEN
+    IF procedure_name = 'sp_updateextendedproperty' OR procedure_name = 'sp_dropextendedproperty' THEN
+      RAISE EXCEPTION 'Property cannot be updated or deleted. Property ''%'' does not exist for ''%''.', "@name", coalesce(var_object_name, 'object specified');
+      RETURN;
+    END IF;
+  ELSE
+    IF procedure_name = 'sp_addextendedproperty' THEN
+      RAISE EXCEPTION 'Property cannot be added. Property ''%'' already exists for ''%''.', "@name", coalesce(var_object_name, 'object specified');
+    END IF;
+  END IF;
+
+  IF procedure_name = 'sp_addextendedproperty' THEN
+    INSERT INTO sys.babelfish_extended_properties(dbid, schema_name, major_name, minor_name, type, name, value) VALUES(var_dbid, var_schema_name, var_major_name, var_minor_name, var_type, "@name", "@value");
+  ELSIF procedure_name = 'sp_updateextendedproperty' THEN
+    UPDATE sys.babelfish_extended_properties SET value = "@value" WHERE dbid = var_dbid AND schema_name = var_schema_name COLLATE "C" AND major_name = var_major_name COLLATE "C" AND minor_name = var_minor_name COLLATE "C" AND type = var_type COLLATE "C" AND name = "@name";
+  ELSIF procedure_name = 'sp_dropextendedproperty' THEN
+    DELETE FROM sys.babelfish_extended_properties WHERE dbid = var_dbid AND schema_name = var_schema_name COLLATE "C" AND major_name = var_major_name COLLATE "C" AND minor_name = var_minor_name COLLATE "C" AND type = var_type COLLATE "C" AND name = "@name";
+  END IF;
+
+  PERFORM sys.babelfish_set_role(var_user);
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE PROCEDURE sys.sp_addextendedproperty
```

Comment:
I wonder if there is any permission check on who could insert / update / delete an extended property?

Reply:
object_id() do the permission check for us. It can return result only when the user has permission on the object.

Reply:
I think you are assuming that if a user could use or execute the object, the he will be able to add a property. 
Because that's how obect_id is implemented. But that's incorrect. 

According to the T-SQL doc: 

Permissions
Members of the db_owner and db_ddladmin fixed database roles can add extended properties to any object with the following exception: db_ddladmin cannot add properties to the database itself, or to users or roles.

Users can add extended properties to objects they own or have ALTER or CONTROL permissions on.

Permissions
Permissions to list extended properties of objects vary by object type.